internal/providers/packet/packet.go

```diff
@@ -0,0 +1,130 @@
+// Copyright 2016 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package packet
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/coreos/coreos-metadata/internal/providers"
+	"github.com/coreos/coreos-metadata/internal/retry"
+)
+
+func FetchMetadata() (providers.Metadata, error) {
+	body, err := retry.Client{
+		InitialBackoff: time.Second,
+		MaxBackoff:     time.Second * 5,
+		MaxAttempts:    10,
+	}.Get(metadataURL)
+	if err != nil {
+		return providers.Metadata{}, err
+	}
+
+	metadata := packetMetadata{}
+
+	if err := json.Unmarshal(body, &metadata); err != nil {
+		return providers.Metadata{}, err
+	}
+
+	publicIPv4, err := metadata.publicIPv4()
+	if err != nil {
+		return providers.Metadata{}, err
+	}
+
+	privateIPv4, err := metadata.privateIPv4()
+	if err != nil {
+		return providers.Metadata{}, err
+	}
+
+	publicIPv6, err := metadata.publicIPv6()
+	if err != nil {
+		return providers.Metadata{}, err
+	}
+
+	return providers.Metadata{
+		Attributes: map[string]string{
+			"PACKET_HOSTNAME":     metadata.Hostname,
+			"PACKET_IPV4_PUBLIC":  providers.String(publicIPv4),
+			"PACKET_IPV4_PRIVATE": providers.String(privateIPv4),
+			"PACKET_IPV6_PUBLIC":  providers.String(publicIPv6),
+		},
+		SshKeys: metadata.SSHKeys,
+	}, nil
+}
+
+const metadataURL = "https://metadata.packet.net/metadata"
+
+var errNotFound = errors.New("Coudn't find an IP with requested attributes")
+
+type packetMetadata struct {
+	Hostname string        `json:"hostname"`
+	Network  packetNetwork `json:"network"`
+	SSHKeys  []string      `json:"ssh_keys"`
+}
+
+type packetNetwork struct {
+	Addresses []packetAddress `json:"addresses"`
+}
+
+type packetAddress struct {
+	Family  int    `json:"address_family"`
+	Address string `json:"address"`
+	Public  bool   `json:"public"`
+}
+
+func (p *packetMetadata) findIP(family int, public bool) (net.IP, error) {
+	for _, addr := range p.Network.Addresses {
+		if addr.Public == public && addr.Family == family {
+			ip := net.ParseIP(addr.Address)
+			if ip == nil {
+				return nil, fmt.Errorf("couldn't parse %q as IP address", addr.Address)
+			}
+
+			return ip, nil
+		}
+	}
+
+	return nil, errNotFound
```

Comment:
why not return nil, nil here, convert the below findIP callers to direct returns of findIP() and remove errNotFound altogether from this listing?

It's not clear to me what value that is adding other than additional verbosity